### PR TITLE
Update conditions to hide duotone panel

### PR DESCRIPTION
**Pull request by oandregal**
Stems from this issue https://github.com/WordPress/gutenberg/pull/33280
Follows up https://github.com/WordPress/gutenberg/pull/32002 and https://github.com/WordPress/gutenberg/issues/31764

This updates the conditions to show/hide the duotone panel. See full panel enabled for reference:

Captura de ecrã de 2021-07-08 17-03-19

- The duotone presets are hidden if `color.duotone` is empty or `null` (it no longer hides the whole duotone component if the `color.duotone` is null):

Captura de ecrã de 2021-07-08 17-09-33

- If `color.customDuotone` is false or ( `color.custom` is false and `color.palette` is empty) we have this:

Captura de ecrã de 2021-07-08 21-33-44

- If only `color.custom` is false we have this:

Captura de ecrã de 2021-07-08 21-34-16

- The shadows/highlight presets are hidden if `color.palette` is empty on null (the control is still visible if `color.custom` is not false):

Captura de ecrã de 2021-07-08 17-02-59

- The duotone control is hidden if `color.duotone` is empty and `color.customDuotone` is false (and also if `color.palette` is empty and `color.custom` is false, independently from what the `color.customDuotone` value is):

Captura de ecrã de 2021-07-08 17-01-11

### CODE REVIEW

**Review by ajlende — CHANGES_REQUESTED** [review]
> it no longer hides the whole duotone component if the `color.duotone` is `null`

I think we should be able to disable duotone without disabling the color palette or custom colors which is why it worked the way it did previously. I'm okay with having `null` (or maybe `false` instead?) disable the panel entirely while an empty array just removes the presets.

EDIT: However, it still makes sense to me to hide the whole thing when all three (`color.custom`, `color.palette`, _and_ `color.duotone`) would result in the empty panel shown in https://github.com/WordPress/gutenberg/pull/33280#issuecomment-876296816.

Captura de ecrã de 2021-07-08 11-44-53

**Comment by oandregal** [comment]
> I think we should be able to disable duotone without disabling the color palette or custom colors which is why it worked the way it did previously. I'm okay with having null (or maybe false instead?) disable the panel entirely while an empty array just removes the presets.

I'm sorry I've come late to the discussion, this is a bit disrupting for everyone. Let me share my mindset, which is to aim for consistency among all the tools consumers have and don't optimize for a specific one.

The duotone panel can work without duotone presets (`color.duotone`): there's still the color palette and the custom colors the user can play with. This is the same situation the other features have: the color panel has colors, gradients, links, and custom; the typography panel has font size, line height, font families; etc.

I understand how useful it can be to have a switch to disable panels entirely, be it duotone, colors, typography, etc. What are our options here?

1. Reusing the same setting for hiding presets or the panel entirely. I'm not a fan of this approach, it adds confusion. While it could be argued that for the duotone panel the `color.duotone` might be a good choice, it gets muddier for the other features: color has 2 different color palettes (`color.palette` and `color.gradients`), typography has font sizes and font families, etc.

2. Create a new setting to hide the panels entirely. I'd favor this one. The way I'd approach this is to land this PR as it is and then discuss if we can get this new setting on time for 5.8. If we don't, then we ship it in 5.9. I presume this is fine as the other panels have been working without a switch to disable them for a few WordPress versions.

I'd like to hear thoughts from others @jorgefilipecosta @youknowriad @mtias @jasmussen

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @nosolosw,

I think duotone should be consistent with the other presets we have. In that case we would have duotone presets that are always an array, and a customDuotone flag. Exactly as it happens on color, gradients, and font sizes.

Duotone appearing or not should not depend on colors. Duotone would not appear if duotone presets is empty and customDuotone is false, and it would appear otherwise.
If colors are disabled (color presets and customColors) but customDuotone is true it would still be possible to pick custom colors for a custom duotone. The same happens on gradients if customColors is false but customGradients is true I can still pick colors for building my gradients. They are independent controls.

**Comment by oandregal** [comment]
So, Jorge is arguing that, in addition to all this PR does, the whole section below should be able to be disabled via a new `color.customDuotone`:

image

**Comment by jorgefilipecosta** [comment]
Exactly,
color.duotone would control this part:
image

And that part would not be available if color.duotone is an empty array.


culor.customDuotone would control this part:
image

And that part would not be available if color.customDuotone is false.


If color.duotone is empty and color.customDuotone is false, the whole duotone functionality would not be available at all. And we would not have any reference to duotone on the UI.

**Comment by ajlende** [comment]
@nosolosw I like the idea of having an option to disable panels that is separate from the preset arrays. I wanted to do something like that for duotone when I was working on it, but ended up going with the established pattern instead to avoid confusion for folks. Seems like something worth pursuing for 5.9.

Adding a separate `color.customDuotone` flag instead of relying on `color.custom` to disable it makes sense to me like @jorgefilipecosta said.

- We'd have `color.duotone` control the top presets.

  Captura de ecrã de 2021-07-08 17-09-33

- And then `color.customDuotone` control the bottom options.

  Captura de ecrã de 2021-07-08 17-25-57

- I think `color.custom` can still disable the custom options.

  Captura de ecrã de 2021-07-08 17-01-31

- And `color.palette` can still disable the palette. (I'd like to hear what @jasmussen has to say, but I think having this set to false can remove the Shadows and Highlights entirely, leaving just the duotone bar. This part probably doesn't have to be changed with this PR.)

  Captura de ecrã de 2021-07-08 17-02-59

- To disable the panel entirely, you'd have to set both `color.duotone` to an empty array and `color.customDuotone` to `false`.

  Captura de ecrã de 2021-07-08 17-01-11

**Comment by oandregal** [comment]
@ajlende nice, I just did that. I think we all three are in agreement.

Now, we need a last little thing: what should happen when `color.custom` is false? At the moment, it removes the "custom color" links plus the horizontal bar between the duotone presets and the shadows/highlights, see:

Captura de ecrã de 2021-07-08 20-55-16

However, I wonder why that bar is controlled by the `color.custom`? As I understand this interaction, that bar should be controlled by `color.customDuotone`: it should be visible if either the presets for shadows/highlights are visible or the custom color links are enabled. Does that make sense to you?

**Comment by ajlende** [comment]
My thinking was that the duotone bar should be controlled by `color.custom` because you can only select custom colors with it. If you don't want to allow custom colors, then it needs to be either disabled or removed.

So the bar should be controlled by both `color.customDuotone` _and_ `color.custom`. Both need to be enabled for it to show.

**Comment by oandregal** [comment]
Ok, I tried to change the logic locally to see how this feels.

This is with `color.custom` false and only controlling the the "custom color" links:

Captura de ecrã de 2021-07-08 21-05-13

And this is with `color.palette` empty:

Captura de ecrã de 2021-07-08 21-08-22

**Comment by oandregal** [comment]
> My thinking was that the duotone bar should be controlled by color.custom because you can only select custom colors with it. If you don't want to allow custom colors, then it needs to be either disabled or removed.

Oh, wait. Can I interact with the bar to modify those colors? I think this may be the reason why I'm confused. If I should be able to do that, we do need to use `color.custom`, yes. The issue is that it doesn't work for me in either chrome or firefox (Ubuntu 21.04) :(

**Comment by ajlende** [comment]
Tested against trunk on Arch Linux Firefox 89.0.2 and Chromium 91.0.4472.114, and custom color picking using the bar is working fine for me both mouse and keyboard only. I don't have Ubuntu ready to go on any other computers or VMs right now. It's strange that it isn't working for you.

**Comment by oandregal** [comment]
OK, pushed all that you mentioned. I think it's ready to ship.

In talking with Jorge about this he raised the concern of using `color.custom` to disable custom duotone colors. It could be argued that it should be replaced by other things (`color.customDuotoneShadows`, etc?). To me, it boils down to this: why someone disables the custom colors for users? Because they want the users to only be able to use a predefined set of colors everywhere, so I think using `color.custom` in the way it's used here is fine.

At the same time it has been a long day for me, so I'd welcome more eyes, including my own rested ones tomorrow :)

File: packages/block-editor/src/hooks/duotone.js [hunk 1]
```diff
@@ -186,15 +199,10 @@ function addDuotoneAttributes( settings ) {
...
-		const hasDuotoneSupport = hasBlockSupport(
-			props.name,
-			'color.__experimentalDuotone'
-		);
-
 		return (
 			<>
 				<BlockEdit { ...props } />
-				{ hasDuotoneSupport && <DuotonePanel { ...props } /> }
+				<DuotonePanel { ...props } />
```
Comment (ajlende) [inline packages/block-editor/src/hooks/duotone.js]: Why did you move this check inside? It seems more natural here and has the added benefit that React doesn't have to bother rendering or running any of the hooks when it's not supported.

File: packages/block-editor/src/hooks/duotone.js [hunk 2]
```diff
@@ -120,19 +122,30 @@ ${ selector } {
...
 	const duotone = style?.color?.duotone;
 
-	const duotonePalette = useSetting( 'color.duotone' );
-	const colorPalette = useSetting( 'color.palette' );
+	const duotonePalette = useSetting( 'color.duotone' ) || EMPTY_ARRAY;
+	const colorPalette = useSetting( 'color.palette' ) || EMPTY_ARRAY;
 	const disableCustomColors = ! useSetting( 'color.custom' );
+	const disableCustomDuotone =
+		! useSetting( 'color.customDuotone' ) ||
+		( colorPalette?.length === 0 && disableCustomColors );
```
Comment (ajlende) [inline packages/block-editor/src/hooks/duotone.js:134]: I would expect this check to go [inside `DuotonePicker`](https://github.com/WordPress/gutenberg/pull/33295/files#diff-37be335b0d340015a545c960c96de07c7e0ea34ce12b3dd8a1735930a526525aR82-R85) where the `colorPalette` is actually used.

**Review by ajlende — COMMENTED** [review]
Seems to be working fine for me. Just had a couple code suggestions before getting this shipped.

File: packages/block-editor/src/hooks/duotone.js (hunk 1 above)
Comment (oandregal) [reply packages/block-editor/src/hooks/duotone.js]: I moved it back to the original position.

I inlined them with the rest while I was working on this to gain a clear picture of the conditions for visibility. Yeah, it's more performant the way it's now. After being more involved with these and with the changes, I'm also confident that the code is readable.

File: packages/block-editor/src/hooks/duotone.js (hunk 2 above)
Comment (oandregal) [reply packages/block-editor/src/hooks/duotone.js:134]: It's a few different things: within the component, we need to make sure the individual subcomponents are visible/hidden when correspond, and we do that. We also need to check that the whole component is hidden when all subparts are also hidden: we need to check for that at this level, before the component is rendered.

**Comment by jasmussen** [comment]
Thanks for your work on this, and I'm happy to see you working out the best path forward.

With regards to supporting custom colors, theme colors, various palettes and such, I don't thin the following note changes anything, but I wanted to point out that as part of https://github.com/WordPress/gutenberg/issues/27331, we'll want a new color swatch picker that groups custom colors and swatches together. This is an older design, but it shows the flow:



The final design should be more like what's shown in https://github.com/WordPress/gutenberg/issues/27473#issuecomment-796773866. But the ultimate goal is to not have to duplicate the entire palette for every color property there is, but instead surface those once you click the property first.

**Comment by oandregal** [comment]
Cool, I've addressed all the comments and added docs as well. Would anyone mind giving a final review to this and green check as appropriate?

I've looked at https://github.com/WordPress/gutenberg/pull/33295#issuecomment-876697570 with fresh eyes and still think the pre-existing logic for `color.custom` makes sense. I also like the addition of `color.customDuotone`. This works consistently with the other features that we have in my view.

**Review by ajlende — APPROVED** [review]
Working well for me 👍
